File: scala/problem14.scala
/*
https://projecteuler.net/problem=14

Longest Collatz sequence
*/

object problem14 {
  private def collatzLength(n: Long): Int = {
    var value = n
    var len = 1
    while (value != 1L) {
      value = if (value % 2L == 0L) value / 2L else 3L * value + 1L
      len += 1
    }
    len
  }

  def baselineAnswer: Long = {
    (1L until 1000000L).maxBy(collatzLength)
  }

  def optimizedAnswer: Long = {
    val limit = 1000000
    val memo = Array.fill(limit)(0)
    memo(1) = 1

    def collatzLengthMemo(n: Long): Int = {
      if (n < limit && memo(n.toInt) != 0) {
        memo(n.toInt)
      } else {
        val next = if ((n & 1L) == 0L) n / 2L else 3L * n + 1L
        val len = 1 + collatzLengthMemo(next)
        if (n < limit) {
          memo(n.toInt) = len
        }
        len
      }
    }

    var bestN = 1L
    var bestLen = 1
    var n = 1L
    while (n < limit) {
      val len = collatzLengthMemo(n)
      if (len > bestLen) {
        bestLen = len
        bestN = n
      }
      n += 1L
    }

    bestN
  }

  case class Result(value: Long, timeNs: Long)

  def answer: Long = {
    optimizedAnswer
  }

  def eulerBaseline(): Result = {
    val start = System.nanoTime()
    val value = baselineAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def euler(): Result = {
    val start = System.nanoTime()
    val value = optimizedAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def main(args: Array[String]): Unit = {
    println("Project Euler Problem 14")

    val baseline = eulerBaseline()
    val optimized = euler()

    println(s"Result (baseline): ${baseline.value}")
    println(s"Time (baseline): ${baseline.timeNs} ns")
    println(s"Result (optimized): ${optimized.value}")
    println(s"Time (optimized): ${optimized.timeNs} ns")

    if (optimized.timeNs > 0) {
      val speedup = baseline.timeNs.toDouble / optimized.timeNs.toDouble
      println(s"Speedup: ${speedup}x")
    }
  }
}

File: cpp/problem14.cpp
/*
https://projecteuler.net/problem=14

Longest Collatz sequence

Which starting number, under one million, produces the longest Collatz chain?
*/

#include <chrono>
#include <iostream>
#include <vector>
using namespace std;

struct result {
    long long value;
    long long time_ns;
};

long long collatz_length_plain(long long n) {
    long long len = 1;
    while (n != 1) {
        if (n % 2 == 0) {
            n /= 2;
        } else {
            n = 3 * n + 1;
        }
        ++len;
    }
    return len;
}

long long evaluate_baseline() {
    long long best_start = 1;
    long long best_len = 1;

    for (long long n = 1; n < 1000000; ++n) {
        long long len = collatz_length_plain(n);
        if (len > best_len) {
            best_len = len;
            best_start = n;
        }
    }

    return best_start;
}

long long collatz_length_memo(long long n, std::vector<int> &memo) {
    if (n < static_cast<long long>(memo.size()) && memo[n] != 0) {
        return memo[n];
    }

    long long next = (n % 2 == 0) ? (n / 2) : (3 * n + 1);
    long long len = 1 + collatz_length_memo(next, memo);

    if (n < static_cast<long long>(memo.size())) {
        memo[n] = static_cast<int>(len);
    }

    return len;
}

long long evaluate_optimized() {
    std::vector<int> memo(1000000, 0);
    memo[1] = 1;

    long long best_start = 1;
    long long best_len = 1;

    for (long long n = 1; n < 1000000; ++n) {
        long long len = collatz_length_memo(n, memo);
        if (len > best_len) {
            best_len = len;
            best_start = n;
        }
    }

    return best_start;
}

result euler() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_optimized();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

result euler_baseline() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_baseline();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

int main() {
    std::cout << "Project Euler Problem 14" << std::endl;

    result baseline = euler_baseline();
    result optimized = euler();

    std::cout << "Result (baseline): " << baseline.value << std::endl;
    std::cout << "Time (baseline): " << baseline.time_ns << " ns" << std::endl;
    std::cout << "Result (optimized): " << optimized.value << std::endl;
    std::cout << "Time (optimized): " << optimized.time_ns << " ns" << std::endl;

    if (optimized.time_ns > 0) {
        double speedup = static_cast<double>(baseline.time_ns) / static_cast<double>(optimized.time_ns);
        std::cout << "Speedup: " << speedup << "x" << std::endl;
    }

    return 0;
}
